app/scripts/metamask-controller.js

```diff
@@ -7798,6 +7802,50 @@ export default class MetamaskController extends EventEmitter {
           .showConfirmationAdvancedDetails;
       },
       getHDEntropyIndex: this.getHDEntropyIndex.bind(this),
+      getNetworkRpcUrl: (chainId) => {
+        // TODO: Move to @metamask/network-controller
+        try {
+          // Find the network client ID for this chain
+          const networkClientId =
+            this.networkController.findNetworkClientIdByChainId(chainId);
+
+          if (networkClientId) {
+            // Get the network configuration
+            const networkConfig =
+              this.networkController.getNetworkConfigurationByNetworkClientId(
+                networkClientId,
+              );
+
+            if (networkConfig) {
+              // Try direct rpcUrl property first
+              if (networkConfig.rpcUrl) {
+                return networkConfig.rpcUrl;
+              }
+
+              // Try rpcEndpoints array
+              if (
+                networkConfig.rpcEndpoints &&
+                networkConfig.rpcEndpoints.length > 0
+              ) {
+                // Use the default endpoint if specified
+                const defaultEndpointIndex =
+                  networkConfig.defaultRpcEndpointIndex || 0;
+                if (networkConfig.rpcEndpoints[defaultEndpointIndex]) {
+                  return networkConfig.rpcEndpoints[defaultEndpointIndex].url;
+                }
+
+                // Fallback to first endpoint
+                if (networkConfig.rpcEndpoints[0].url) {
+                  return networkConfig.rpcEndpoints[0].url;
+                }
+              }
+            }
+          }
+        } catch (error) {
+          console.error('Error getting RPC URL:', error);
+        }
+        return 'unknown';
```

Comment:
small nit: It should be safe to trust the return types here. If the type indicates that value will not return `undefined`/`null`/etc you should be able to trust it (if you found that we can't trust it, either the code or the types are broken and should be fixed or a new issue created to fix them).

We _probably_ don't need all of the falsey checks (`if (networkClientId)`, `if (networkConfig)`), , but I haven't checked all of these return types myself.

There are cases when we must should guard against unexpected data (like we do in migration files, and to a much larger degree), I'm not sure if this is one of those cases (I could be wrong though!)

Reply:
ok, I've updated I will do some more tests now to ensure everything is working. But looking at the code closer and types. I think it's safe.